cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1897,7 +1897,8 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
 
     auto handle_recursion = [&](int32_t nesting_depth) -> Result<bool> {
       RETURN_NOT_OK(CheckNestingDepth(nesting_depth));
-      return select.recursive && nesting_depth <= select.max_recursion;
+      return select.type == SelectorType::Directory &&
```

Comment:
Is it possible to reach this point and `select.type != SelectorType::Directory`?